pkg/spdk/server.go

```diff
@@ -346,18 +346,145 @@ func (s *Server) ReplicaSnapshotDelete(ctx context.Context, req *spdkrpc.Snapsho
+	if req.Name == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "replica name is required")
+	}
+	if req.DstReplicaName == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "dst replica name is required")
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	r := s.replicaMap[req.Name]
+
+	if r == nil {
+		return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find replica %s during rebuilding src finish", req.Name)
+	}
+
+	if err = r.RebuildingSrcFinish(s.spdkClient, req.DstReplicaName); err != nil {
+		return nil, err
+	}
+	return &empty.Empty{}, nil
+}
+
+func (s *Server) ReplicaSnapshotShallowCopy(ctx context.Context, req *spdkrpc.ReplicaSnapshotShallowCopyRequest) (ret *empty.Empty, err error) {
+	if req.Name == "" || req.SnapshotName == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "replica snapshot name is required")
+	}
+
+	s.RLock()
+	r := s.replicaMap[req.Name]
+	defer s.RUnlock()
+
+	if r == nil {
+		return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find replica %s during snapshot %s shallow copy", req.Name, req.SnapshotName)
+	}
+
+	// Cannot add a lock to protect this now since a shallow copy may be time-consuming
+	if err = r.SnapshotShallowCopy(s.spdkClient, req.SnapshotName); err != nil {
+		return nil, err
+	}
+	return &empty.Empty{}, nil
+}
+
+func (s *Server) ReplicaRebuildingDstStart(ctx context.Context, req *spdkrpc.ReplicaRebuildingDstStartRequest) (ret *spdkrpc.ReplicaRebuildingDstStartResponse, err error) {
+	if req.Name == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "replica name is required")
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	r := s.replicaMap[req.Name]
+
+	if r == nil {
+		return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find replica %s during rebuilding dst start", req.Name)
+	}
+
+	address, err := r.RebuildingDstStart(s.spdkClient, req.ExposeRequired)
+	if err != nil {
+		return nil, err
+	}
+	return &spdkrpc.ReplicaRebuildingDstStartResponse{Address: address}, nil
+}
+
+func (s *Server) ReplicaRebuildingDstFinish(ctx context.Context, req *spdkrpc.ReplicaRebuildingDstFinishRequest) (ret *empty.Empty, err error) {
+	if req.Name == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "replica name is required")
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	r := s.replicaMap[req.Name]
+
+	if r == nil {
+		return nil, grpcstatus.Errorf(grpccodes.NotFound, "cannot find replica %s during rebuilding dst finish", req.Name)
+	}
+
+	if err = r.RebuildingDstFinish(s.spdkClient, req.UnexposeRequired); err != nil {
+		return nil, err
+	}
+	return &empty.Empty{}, nil
+}
+
+func (s *Server) ReplicaRebuildingDstSnapshotCreate(ctx context.Context, req *spdkrpc.SnapshotRequest) (ret *empty.Empty, err error) {
+	if req.Name == "" || req.SnapshotName == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "replica name and snapshot name are required")
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	r := s.replicaMap[req.Name]
+
```

Comment:
Can we use a map lock for reading/updating replicaMap and engineMap?
```
	s.mapLock.RLock()
	r := s.replicaMap[req.Name]
	s.mapLock.RUnlock()
```

Same question for other server's methods.